recommendations.md

```diff
@@ -60,3 +60,204 @@ On the catalog, this means that you should, where possible:
         runAsUser: 0 # root uid == 0
         privileged: true
   ```
+
+## Don't use interpolation in shell scripts
+
+Using `$(tekton.task)` interpolation in the `script` or as a `sh -c`
+string is extremely fragile.  The interpolation done by tekton is not
+aware of the context in which the interpolation happens.  A space, a
+quote sign, a backslash or newline could easily thwart an otherwise
+beautiful script.
+
+```
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.one)
+```
+
+If `params.one` happens to contain a quote, then the resultin shell
+script might look like this:
+
+```
+echo '
+```
+
+This script is not valid, and the task will fail.
```

Comment:
I think :

```bash
var=$(cat <<EOF
$(params.one)
EOF
)
```

should work too

Reply:
Almost.  Shell escape rules are treacherous:
* If I send in `$(cat)` as the parameter, the script will evaluate `cat` which will likely fail because there is nothing in stdin
* If I send in "EOF" then the parameter will fail, and have the value EOF.
I'm pretty sure that there are no "good" interpolation rules for shell (or any language for that matter); because tekton just doesn't know (and rightly so) about the chosen language's escaping rules.

So in my opinion, the rule of thumb should be to avoid it unless you're very certain about the data being passed in.

Reply:
that's right 👍🏽